<a href="https://colab.research.google.com/github/athifashihab/FDS/blob/main/notebooks/class%20activities/6_AirQuality_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Group No: 6
## Members:
- ATHIFA SHARIN SHIHAB ERAMANGALATH(U24200842)
- AISHA JALIL IBRAHIM AL DAHHAN(U25103468)
- KHULOOD ISMAIL HAJ ABED GHAZAL(U25200155)

**In-Class Team-Based Learning Activity** : Analyzing California Air Quality Data **[15 MARKS]**
- Objective:
    - In this activity, YOU will work in teams to analyze the California Air Quality 2020 dataset. Download it from the following link:
    [California Air Quality Data](https://www.kaggle.com/datasets/thaddeussegura/california-air-quality-2020-through-sept10th)

    -  columns:
      - `Date`: The date of the measurement.
      - `Daily Mean PM2.5 Concentration`: Average PM2.5 concentration for the day.
      - `AQI`: Air Quality Index derived from PM2.5 levels.
      - `CBSA Name`: Core-Based Statistical Area (e.g., Fresno, Riverside).
      - `County Name`: County where the measurement was taken.
      - Other metadata like `latitude`, `longitude`, and units.
    
    - You will:
      - Explore the dataset to understand its structure.
      - Perform operations such as sorting, filtering, and selecting rows to extract meaningful insights.
      - Collaborate to answer questions about air quality trends and their implications.


In [ ]:
!pip install datascience

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 725.1/725.1 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 50.7 MB/s eta 0:00:00


In [ ]:
from datascience import *
import numpy as np

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving California_airquality.csv to California_airquality.csv


**Step-by-Step Instructions**
- **Step 1**: Introduction to the Dataset
    - Goal : Familiarize yourself with the dataset and its columns.
      - **Task 1.1**: Load the dataset into a `table` and display the first few rows.

In [ ]:
# Task 1.1: Load the dataset into a Table and display the first few rows
path_data = ''
airquality = Table.read_table(path_data + 'California_airquality.csv')

airquality.show(5)

Date,Source,Site ID,POC,Daily Mean PM2.5 Concentration,UNITS,DAILY_AQI_VALUE,Site Name,DAILY_OBS_COUNT,PERCENT_COMPLETE,AQS_PARAMETER_CODE,AQS_PARAMETER_DESC,CBSA_CODE,CBSA_NAME,STATE_CODE,STATE,COUNTY_CODE,COUNTY,SITE_LATITUDE,SITE_LONGITUDE
1/1/20,AQS,60010007,3,8.6,ug/m3 LC,36,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
1/2/20,AQS,60010007,3,4.5,ug/m3 LC,19,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
1/3/20,AQS,60010007,3,14.2,ug/m3 LC,55,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
1/4/20,AQS,60010007,3,10.9,ug/m3 LC,45,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
1/5/20,AQS,60010007,3,7.8,ug/m3 LC,33,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784


**Step 2**: Sorting Rows
  - Goal : Teach students how to sort rows based on specific criteria.
    - **Task 2.1** : Sort the dataset by Daily Mean PM2.5 Concentration in descending order to find the days with the worst air quality.

In [ ]:
# Task 2.1: Sort by Daily Mean PM2.5 Concentration in descending order
sorted_by_pm25 = airquality.sort('Daily Mean PM2.5 Concentration', descending=True)
sorted_by_pm25.show(10)

Date,Source,Site ID,POC,Daily Mean PM2.5 Concentration,UNITS,DAILY_AQI_VALUE,Site Name,DAILY_OBS_COUNT,PERCENT_COMPLETE,AQS_PARAMETER_CODE,AQS_PARAMETER_DESC,CBSA_CODE,CBSA_NAME,STATE_CODE,STATE,COUNTY_CODE,COUNTY,SITE_LATITUDE,SITE_LONGITUDE
8/21/20,AirNow,60871005,3,378.5,ug/m3 LC,419,San Lorenzo Valley Middle School,1,100,88101,PM2.5 - Local Conditions,42100,"Santa Cruz-Watsonville, CA",6,California,87,Santa Cruz,37.0632,-122.083
8/20/20,AirNow,60798002,3,304.1,ug/m3 LC,354,Atascadero,1,100,88101,PM2.5 - Local Conditions,42020,"San Luis Obispo-Paso Robles-Arroyo Grande, CA",6,California,79,San Luis Obispo,35.4945,-120.666
8/29/20,AQS,60631006,4,276.8,ug/m3 LC,327,Quincy-N Church Street,1,100,88502,Acceptable PM2.5 AQI & Speciation Mass,nan,nan,6,California,63,Plumas,39.9396,-120.944
9/10/20,AirNow,60431001,3,263.3,ug/m3 LC,314,Yosemite NP-Yosemite Village Vistor Center,1,100,88502,Acceptable PM2.5 AQI & Speciation Mass,nan,nan,6,California,43,Mariposa,37.7487,-119.587
8/25/20,AQS,60631006,4,251.5,ug/m3 LC,302,Quincy-N Church Street,1,100,88502,Acceptable PM2.5 AQI & Speciation Mass,nan,nan,6,California,63,Plumas,39.9396,-120.944
8/20/20,AirNow,60871005,3,243.9,ug/m3 LC,294,San Lorenzo Valley Middle School,1,100,88101,PM2.5 - Local Conditions,42100,"Santa Cruz-Watsonville, CA",6,California,87,Santa Cruz,37.0632,-122.083
8/21/20,AirNow,60798002,3,236.2,ug/m3 LC,286,Atascadero,1,100,88101,PM2.5 - Local Conditions,42020,"San Luis Obispo-Paso Robles-Arroyo Grande, CA",6,California,79,San Luis Obispo,35.4945,-120.666
8/24/20,AQS,60631006,4,232.5,ug/m3 LC,282,Quincy-N Church Street,1,100,88502,Acceptable PM2.5 AQI & Speciation Mass,nan,nan,6,California,63,Plumas,39.9396,-120.944
8/26/20,AQS,60631006,4,231.7,ug/m3 LC,281,Quincy-N Church Street,1,100,88502,Acceptable PM2.5 AQI & Speciation Mass,nan,nan,6,California,63,Plumas,39.9396,-120.944
8/30/20,AQS,60631006,4,222,ug/m3 LC,272,Quincy-N Church Street,1,100,88502,Acceptable PM2.5 AQI & Speciation Mass,nan,nan,6,California,63,Plumas,39.9396,-120.944


In [ ]:
# To understand the sorting more clearly
sorted_by_pm25.select('Date', 'Daily Mean PM2.5 Concentration', 'CBSA_NAME', 'COUNTY').show(10)

Date,Daily Mean PM2.5 Concentration,CBSA_NAME,COUNTY
8/21/20,378.5,"Santa Cruz-Watsonville, CA",Santa Cruz
8/20/20,304.1,"San Luis Obispo-Paso Robles-Arroyo Grande, CA",San Luis Obispo
8/29/20,276.8,nan,Plumas
9/10/20,263.3,nan,Mariposa
8/25/20,251.5,nan,Plumas
8/20/20,243.9,"Santa Cruz-Watsonville, CA",Santa Cruz
8/21/20,236.2,"San Luis Obispo-Paso Robles-Arroyo Grande, CA",San Luis Obispo
8/24/20,232.5,nan,Plumas
8/26/20,231.7,nan,Plumas
8/30/20,222,nan,Plumas


**Teams should discuss why certain days might have had poor air quality (e.g., wildfires)!**

> **Discussion:** The days with the highest PM2.5 concentrations fall in August–September 2020. This aligns with California's record-breaking 2020 wildfire season. Wildfires release massive amounts of fine particulate matter (PM2.5) into the air, dramatically worsening air quality across wide regions. Counties like Santa Cruz, San Luis Obispo, Plumas, and Mariposa were particularly hard hit.


- **Task 2.2** : Sort the dataset alphabetically by `CBSA_NAME` to group data by region.

In [ ]:
# Task 2.2: Sort alphabetically by CBSA_NAME
sorted_by_cbsa = airquality.sort('CBSA_NAME')
sorted_by_cbsa.show(10)

Date,Source,Site ID,POC,Daily Mean PM2.5 Concentration,UNITS,DAILY_AQI_VALUE,Site Name,DAILY_OBS_COUNT,PERCENT_COMPLETE,AQS_PARAMETER_CODE,AQS_PARAMETER_DESC,CBSA_CODE,CBSA_NAME,STATE_CODE,STATE,COUNTY_CODE,COUNTY,SITE_LATITUDE,SITE_LONGITUDE
1/1/20,AQS,60290010,1,36.8,ug/m3 LC,104,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/4/20,AQS,60290010,1,23.7,ug/m3 LC,75,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/10/20,AQS,60290010,1,18.5,ug/m3 LC,64,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/13/20,AQS,60290010,1,15.1,ug/m3 LC,57,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/16/20,AQS,60290010,1,21,ug/m3 LC,70,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/19/20,AQS,60290010,1,23.9,ug/m3 LC,76,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/22/20,AQS,60290010,1,11.9,ug/m3 LC,50,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/25/20,AQS,60290010,1,20.6,ug/m3 LC,69,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/28/20,AQS,60290010,1,20.1,ug/m3 LC,68,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015
1/31/20,AQS,60290010,1,22.4,ug/m3 LC,73,Bakersfield-Golden/M St,1,100,88101,PM2.5 - Local Conditions,12540,"Bakersfield, CA",6,California,29,Kern,35.3856,-119.015


In [ ]:
# To understand the sorting more clearly
sorted_by_cbsa.select('Date', 'CBSA_NAME', 'Daily Mean PM2.5 Concentration').show(10)

Date,CBSA_NAME,Daily Mean PM2.5 Concentration
1/1/20,"Bakersfield, CA",36.8
1/4/20,"Bakersfield, CA",23.7
1/10/20,"Bakersfield, CA",18.5
1/13/20,"Bakersfield, CA",15.1
1/16/20,"Bakersfield, CA",21
1/19/20,"Bakersfield, CA",23.9
1/22/20,"Bakersfield, CA",11.9
1/25/20,"Bakersfield, CA",20.6
1/28/20,"Bakersfield, CA",20.1
1/31/20,"Bakersfield, CA",22.4


**Step 3**: Selecting Rows
  - Goal : Practice filtering rows based on specific conditions.
      - **Task 3.1** : Extract rows where the Daily Mean PM2.5 Concentration exceeds 50 µg/m³.

In [ ]:
# Task 3.1: Filter rows where PM2.5 > 50 µg/m³
high_pm25 = airquality.where('Daily Mean PM2.5 Concentration', are.above(50))
print(f'Number of days with PM2.5 > 50 µg/m³: {high_pm25.num_rows}')
high_pm25.show(5)

Number of days with PM2.5 > 50 µg/m³: 354


Date,Source,Site ID,POC,Daily Mean PM2.5 Concentration,UNITS,DAILY_AQI_VALUE,Site Name,DAILY_OBS_COUNT,PERCENT_COMPLETE,AQS_PARAMETER_CODE,AQS_PARAMETER_DESC,CBSA_CODE,CBSA_NAME,STATE_CODE,STATE,COUNTY_CODE,COUNTY,SITE_LATITUDE,SITE_LONGITUDE
8/20/20,AirNow,60010007,3,63.4,ug/m3 LC,155,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
8/21/20,AirNow,60010007,3,54.5,ug/m3 LC,148,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
8/23/20,AirNow,60010007,3,99.9,ug/m3 LC,174,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
8/24/20,AirNow,60010007,3,57.9,ug/m3 LC,152,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784
9/10/20,AirNow,60010007,3,53.8,ug/m3 LC,146,Livermore,1,100,88101,PM2.5 - Local Conditions,41860,"San Francisco-Oakland-Hayward, CA",6,California,1,Alameda,37.6875,-121.784


In [ ]:
# To understand the selection more clearly
high_pm25.select('Date', 'Daily Mean PM2.5 Concentration', 'CBSA_NAME', 'COUNTY').show(5)

Date,Daily Mean PM2.5 Concentration,CBSA_NAME,COUNTY
8/20/20,63.4,"San Francisco-Oakland-Hayward, CA",Alameda
8/21/20,54.5,"San Francisco-Oakland-Hayward, CA",Alameda
8/23/20,99.9,"San Francisco-Oakland-Hayward, CA",Alameda
8/24/20,57.9,"San Francisco-Oakland-Hayward, CA",Alameda
9/10/20,53.8,"San Francisco-Oakland-Hayward, CA",Alameda


**Teams should count how many such days occurred and discuss the health implications.**

> **Discussion:** A total of **354 days** recorded PM2.5 concentrations above 50 µg/m³ across the dataset. PM2.5 > 35 µg/m³ is "Unhealthy" for all people per EPA guidelines. At these levels, everyone may experience health effects: respiratory irritation, reduced lung function, and increased risk of cardiovascular events. Sensitive groups (children, elderly, people with lung/heart conditions) are especially at risk. Most of these extreme readings occurred during the August–September 2020 wildfire events.


- **Task 3.2** : Find all rows corresponding to `Fresno, CA`.

In [ ]:
# Task 3.2: Find all rows for Fresno, CA
fresno = airquality.where('CBSA_NAME', are.equal_to('Fresno, CA'))
print(f'Number of rows for Fresno, CA: {fresno.num_rows}')
fresno.show(5)

Number of rows for Fresno, CA: 1716


Date,Source,Site ID,POC,Daily Mean PM2.5 Concentration,UNITS,DAILY_AQI_VALUE,Site Name,DAILY_OBS_COUNT,PERCENT_COMPLETE,AQS_PARAMETER_CODE,AQS_PARAMETER_DESC,CBSA_CODE,CBSA_NAME,STATE_CODE,STATE,COUNTY_CODE,COUNTY,SITE_LATITUDE,SITE_LONGITUDE
1/1/20,AQS,60190011,1,42.3,ug/m3 LC,118,Fresno - Garland,1,100,88101,PM2.5 - Local Conditions,23420,"Fresno, CA",6,California,19,Fresno,36.7854,-119.773
1/2/20,AQS,60190011,1,35.9,ug/m3 LC,102,Fresno - Garland,1,100,88101,PM2.5 - Local Conditions,23420,"Fresno, CA",6,California,19,Fresno,36.7854,-119.773
1/3/20,AQS,60190011,1,29.6,ug/m3 LC,88,Fresno - Garland,1,100,88101,PM2.5 - Local Conditions,23420,"Fresno, CA",6,California,19,Fresno,36.7854,-119.773
1/4/20,AQS,60190011,1,9.4,ug/m3 LC,39,Fresno - Garland,1,100,88502,Acceptable PM2.5 AQI & Speciation Mass,23420,"Fresno, CA",6,California,19,Fresno,36.7854,-119.773
1/5/20,AQS,60190011,1,13.8,ug/m3 LC,55,Fresno - Garland,1,100,88101,PM2.5 - Local Conditions,23420,"Fresno, CA",6,California,19,Fresno,36.7854,-119.773


In [ ]:
# To understand more clearly
fresno.select('Date', 'Daily Mean PM2.5 Concentration', 'CBSA_NAME').show(5)

Date,Daily Mean PM2.5 Concentration,CBSA_NAME
1/1/20,42.3,"Fresno, CA"
1/2/20,35.9,"Fresno, CA"
1/3/20,29.6,"Fresno, CA"
1/4/20,9.4,"Fresno, CA"
1/5/20,13.8,"Fresno, CA"


**Teams should calculate the average PM2.5 concentration for Fresno during this period.**

In [ ]:
# Calculate average PM2.5 for Fresno
fresno_avg = np.mean(fresno.column('Daily Mean PM2.5 Concentration'))
print(f'Average PM2.5 concentration for Fresno, CA: {fresno_avg:.2f} µg/m³')

Average PM2.5 concentration for Fresno, CA: 11.60 µg/m³


An average PM2.5 concentration of 11.60 µg/m³ in Fresno falls within the “Good” air quality range (below 12 µg/m³).Even though the average is “Good,” it may still hide short periods of higher pollution, so residents could still experience temporary unhealthy air days.

**Step 4**: Creating New Columns
- Goal : Teach students how to compute new metrics and add them to the table
    - **Task 4.1** : Add a column indicating whether the air quality is `"Good" (PM2.5 < 12)`, `"Moderate" (12 ≤ PM2.5 < 35)`, or `"Unhealthy" (PM2.5 ≥ 35)`.

In [ ]:
# Task 4.1: Add 'Air Quality Category' column
def classify_air_quality(pm25):
    if pm25 < 12:
        return 'Good'
    elif pm25 < 35:
        return 'Moderate'
    else:
        return 'Unhealthy'

categories = airquality.apply(classify_air_quality, 'Daily Mean PM2.5 Concentration')
airquality_with_cat = airquality.with_column('Air Quality Category', categories)
airquality_with_cat.select('Date', 'Daily Mean PM2.5 Concentration', 'Air Quality Category').show(5)

Date,Daily Mean PM2.5 Concentration,Air Quality Category
1/1/20,8.6,Good
1/2/20,4.5,Good
1/3/20,14.2,Moderate
1/4/20,10.9,Good
1/5/20,7.8,Good


**Task 4.2** : Calculate the percentage of days in each category for Fresno.

In [ ]:
# Task 4.2: Calculate percentage of days in each category for Fresno
fresno_cat = airquality_with_cat.where('CBSA_NAME', are.equal_to('Fresno, CA'))
total_fresno = fresno_cat.num_rows

good_count      = fresno_cat.where('Air Quality Category', are.equal_to('Good')).num_rows
moderate_count  = fresno_cat.where('Air Quality Category', are.equal_to('Moderate')).num_rows
unhealthy_count = fresno_cat.where('Air Quality Category', are.equal_to('Unhealthy')).num_rows

print('Air Quality Category percentages for Fresno, CA:\n')
print(f'Good:       {good_count / total_fresno * 100:.2f}%')
print(f'Moderate:   {moderate_count / total_fresno * 100:.2f}%')
print(f'Unhealthy:   {unhealthy_count / total_fresno * 100:.2f}%')

Air Quality Category percentages for Fresno, CA:

Good:       73.37%
Moderate:   22.26%
Unhealthy:   4.37%


**Step 5**: Group Challenge
- Goal :  solve a complex problem.
    - **Challenge** : Identify the top 3 regions (CBSA Names) with the highest average PM2.5 concentrations

In [ ]:
# Challenge: Identify the top 3 regions (CBSA Names) with the highest average PM2.5

top3_regions = (
    airquality
    .select('CBSA_NAME', 'Daily Mean PM2.5 Concentration')  # keep only needed columns
    .group('CBSA_NAME', np.mean)                            # group by region, compute mean
    .sort('Daily Mean PM2.5 Concentration mean', descending=True)  # highest first
    .take(np.arange(3))                                     # take top 3 rows
)

top3_regions.show()

CBSA_NAME,Daily Mean PM2.5 Concentration mean
"Hanford-Corcoran, CA",18.794
"Visalia-Porterville, CA",14.5251
"Madera, CA",12.5733


**Grading Distribution (Out of 15)**
- Below is the grading rubric for the tasks in the activity.
    - Each task will be graded based on correctness, completeness, and clarity of the solution.


# **Submission Instructions**


- **1. File Requirements**
  - **File Name** : Save your Jupyter Notebook as Group_No_AirQuality_Analysis.ipynb. Replace Group_No with your assigned group no.
  - **Self-Contained** : Ensure the notebook is self-contained, meaning it includes all necessary code, outputs, and explanations. Use Markdown cells to describe your steps and findings.
  - **Colab Compatibility** : Test that your notebook can be opened and run using "Open in Colab" by including a link to your GitHub repository in the notebook's first cell:

    - [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/your_username/your_repo/blob/main/Group_Name_AirQuality_Analysis.ipynb)
      - Replace your_username, your_repo, and Group_No with your actual GitHub username, repository name, and group no.
- **2. Group Information**
Include the group no and member names at the top of the notebook in a Markdown cell:

# Group No: 6
## Members:
- ATHIFA SHARIN SHIHAB ERAMANGALATH(U24200842)
- AISHA JALIL IBRAHIM AL DAHHAN(U25103468)
- KHULOOD ISMAIL HAJ ABED GHAZAL(U25200155)



- **3. Submission Process**
  - GitHub Repository :
    - Create a public GitHub repository for your group (if not already created)
    - Upload your completed Jupyter Notebook (Group_No_AirQuality_Analysis.ipynb) to the repository.
    - Ensure the repository contains a README.md file with the following:
      - A brief description of the project.
      - Instructions for opening the notebook in Colab.
  - Blackboard Submission :
    - Submit the link to your GitHub repository on Blackboard.
    - Additionally, upload the .ipynb file directly to Blackboard as a backup.
